File: critical-point/core/src/template/equipment.rs
use crate::template::attribute::TmplAttribute;
use crate::template::base::impl_tmpl;
use crate::utils::{impl_for, rkyv_self, JewelSlots, LevelRange, PiecePlus, Table, TmplID};

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
pub enum TmplEquipmentSlot {
    Slot1 = 1,
    Slot2 = 2,
    Slot3 = 3,
}

rkyv_self!(TmplEquipmentSlot);

#[derive(Debug, serde::Serialize, serde::Deserialize, rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
#[rkyv(derive(Debug))]
pub struct TmplEquipment {
    pub id: TmplID,
    pub name: String,
    pub character: TmplID,
    pub slot: TmplEquipmentSlot,
    #[serde(default)]
    pub parents: Table<TmplID, u32>,
    pub level: LevelRange,
    #[serde(default)]
    pub materials: Table<TmplID, Vec<f32>>,
    pub attributes: Table<TmplAttribute, Vec<f32>>,
    #[serde(default)]
    pub slots: Vec<JewelSlots>,
    #[serde(default)]
    pub entries: Table<TmplID, Vec<PiecePlus>>,
    // #[serde(default)]
    // pub script: Option<TmplScript>,
    // #[serde(default)]
    // pub script_args: Table<Symbol, f32>,
}

impl_tmpl!(TmplEquipment, Equipment, "Equipment");

impl_for!(TmplEquipment, ArchivedTmplEquipment, {
    #[inline]
    pub fn level_to_index(&self, level: u32) -> usize {
        (level.clamp(self.level.min, self.level.max) - self.level.min) as usize
    }
});

#[cfg(test)]
mod tests {
    use super::*;
    use crate::template::database::TmplDatabase;
    use crate::utils::id;

    #[test]
    fn test_load_equipment() {
        let db = TmplDatabase::new(10240, 150).unwrap();

        let equipment = db.find_as::<TmplEquipment>(id!("Equipment.No1")).unwrap();
        assert_eq!(equipment.id, id!("Equipment.No1"));
        assert_eq!(equipment.name, "Weapon No1");
        assert_eq!(equipment.character, id!("Character.One"));
        assert_eq!(equipment.slot, TmplEquipmentSlot::Slot1);
        assert_eq!(equipment.level, [1, 4].into());

        assert_eq!(
            equipment.attributes.keys().copied().collect::<Vec<TmplAttribute>>(),
            &[
                TmplAttribute::PhysicalAttack,
                TmplAttribute::ElementalAttack,
                TmplAttribute::ArcaneAttack,
                TmplAttribute::CriticalChance,
            ]
        );
        assert_eq!(equipment.attributes.value_x(0).as_slice(), &[13.0, 19.0, 25.0, 31.0]);
        assert_eq!(equipment.attributes.value_x(1).as_slice(), &[8.0, 12.0, 16.0, 20.0]);
        assert_eq!(equipment.attributes.value_x(2).as_slice(), &[13.0, 18.0, 23.0, 28.0]);
        assert_eq!(equipment.attributes.value_x(3).as_slice(), &[0.02, 0.03, 0.04, 0.05]);

        assert_eq!(
            equipment.slots.as_slice(),
            &[
                JewelSlots::new(0, 0, 0),
                JewelSlots::new(0, 0, 0),
                JewelSlots::new(0, 1, 0),
                JewelSlots::new(0, 1, 0)
            ]
        );

        assert_eq!(equipment.entries.len(), 1);
        assert_eq!(equipment.entries[0].k, id!("Entry.AttackUp"));
        assert_eq!(
            equipment.entries[0].v.as_slice(),
            &[
                PiecePlus::new(1, 0),
                PiecePlus::new(1, 1),
                PiecePlus::new(1, 2),
                PiecePlus::new(1, 3)
            ]
        );

        // assert_eq!(
        //     equipment.script_args.key_iter().cloned().collect::<Vec<Symbol>>(),
        //     &[id!("extra_def"),]
        // );
        // assert_eq!(
        //     equipment
        //         .script_args
        //         .values_iter()
        //         .map(|x| x.to_vec())
        //         .collect::<Vec<Vec<f64>>>(),
        //     &[vec![5.0, 10.0, 15.0, 20.0]]
        // );
    }
}

File: turning-point/tests/equipment.py
from turning_point import *

Equipment(
    'Equipment.No1',
    name="No1",
    icon='icon',
    character='Character.No1',
    position=Position1,
    level=[1, 4],
    materials=[],
    attributes={
        PhysicalAttack: [13, 19, 25, 31],
        ElementalAttack: [8, 12, 16, 20],
        ArcaneAttack: [13, 18, 23, 28],
        CriticalChance: ['2%', '3%', '4%', '5%'],
    },
    slots=['', '', 'A1', 'A1'],
    entries={
        "Entry.AttackUp": [(1,0), (1,1), (1,2), (1,3)],
    },
    script_args={
        "extra_def": [5, 10, 15, 20],
    },
    script="""
    on_assemble {
        secondary.final_skill_damage_ratio *= 110%
    }

    after_assemble {
        extra.cut_defense += A.extra_def
        extra.blunt_defense += A.extra_def
        extra.ammo_defense += A.extra_def
    }
    """
)

Equipment(
    'Equipment.No2',
    name="No2",
    icon='icon',
    character='Character.No1',
    position=Position3,
    level=[0, 3],
    materials=[],
    attributes={
        PhysicalAttack: [10, 15, 20, 25],
        ElementalAttack: [7, 10, 13, 16],
        ArcaneAttack: [8, 12, 16, 20],
        CriticalDamage: ['10%', '12%', '15%', '18%'],
    },
    slots=['A1', 'A2', 'S1A1', 'S1A2'],
    entries={
        "Entry.DefenseUp": [(2,0), (2,1), (2,2), (2,3)],
    },
)

Equipment(
    'Equipment.No3',
    name="No3",
    icon='icon',
    character='Character.No1',
    position=Position1,
    level=[0, 3],
    materials=[],
    attributes={},
)

Equipment(
    'Equipment.No4',
    name="No4",
    icon='icon',
    character='Character.No2',
    position=Position1,
    level=[0, 3],
    materials=[],
    attributes={},
)
